### PR TITLE
feat: add generic store interface and impl

### PR DESCRIPTION
**Pull request by frrist**
This PR implements a `CborIpldStore` interface and `CborIpldBlockstore` interface. Packages such as go-amt-ipld and go-hamt-ipld can then use the `CborIpldStore` interface as a canolical means of accessing their stores, see https://github.com/filecoin-project/go-amt-ipld/pull/8 and https://github.com/ipfs/go-hamt-ipld/pull/4

### CODE REVIEW

File: store.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+	"context"
+	"fmt"
+
+	block "github.com/ipfs/go-block-format"
+	recbor "github.com/polydawn/refmt/cbor"
+	atlas "github.com/polydawn/refmt/obj/atlas"
+
+	cid "github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	cbg "github.com/whyrusleeping/cbor-gen"
```
Comment (anorth) [inline store.go]: nit: import grouping

File: store.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+func (se SerializationError) Error() string {
+	return se.err.Error()
+}
+
+func (se SerializationError) Unwrap() error {
+	return se.err
+}
+
+func (se SerializationError) Is(o error) bool {
```
Comment (anorth) [inline store.go]: This should be a free function, not a method.

File: store.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+}
+
+type cidProvider interface {
+	Cid() cid.Cid
+}
+
+func (s *BasicCborIpldStore) Put(ctx context.Context, v interface{}) (cid.Cid, error) {
+	mhType := uint64(mh.BLAKE2B_MIN + 31)
+	mhLen := -1
+	codec := uint64(cid.DagCBOR)
```
Comment (Kubuxu) [inline store.go]: nit, use `cid.Prefix` from the start (if possible)

File: store.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+	var expCid cid.Cid
+	if c, ok := v.(cidProvider); ok {
+		pref := c.Cid().Prefix()
+		mhType = pref.MhType
+		mhLen = pref.MhLength
+		codec = pref.Codec
+		expCid = c.Cid()
```
Comment (Kubuxu) [inline store.go]: Call `c.Cid()` only once, no need to hash things twice.

File: store.go [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+func (se SerializationError) Error() string {
+	return se.err.Error()
+}
+
+func (se SerializationError) Unwrap() error {
+	return se.err
+}
+
+func IsSerializationError(o error) bool {
```
Comment (Kubuxu) [inline store.go]: I disagree with previous review. It should be a method on the `SerializationError.Is` to support https://godoc.org/errors#Is

File: store.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+		blk, err := block.NewBlockWithCid(buf.Bytes(), c)
+		if err != nil {
+			return cid.Undef, err
+		}
+
+		if err := s.Blocks.Put(blk); err != nil {
+			return cid.Undef, err
+		}
+
+		return blk.Cid(), nil
```
Comment (Kubuxu) [inline store.go]: `expCid` check is not done if object is `cbg.CBORMarshaler`

File: store.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+	block "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	recbor "github.com/polydawn/refmt/cbor"
+	atlas "github.com/polydawn/refmt/obj/atlas"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+type CborIpldStore interface {
```
Comment (Kubuxu) [inline store.go]: Name it just `IpldStore`, `cbor` is already contained in package name.

File: store.go (hunk 3 above)
Comment (frrist) [reply store.go]: I think we could only do that if there some a guarantee that `v` had a `Cid()` method, unless I am misunderstanding something?

File: store.go (hunk 6 above)
Comment (frrist) [reply store.go]: `NewBlockWithCid` has an internal check for this, is it necessary to do it here too?

File: store.go (hunk 7 above)
Comment (frrist) [reply store.go]: will also rename to IpldBLockstore

File: store.go [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+	block "github.com/ipfs/go-block-format"
+	cid "github.com/ipfs/go-cid"
+	mh "github.com/multiformats/go-multihash"
+	recbor "github.com/polydawn/refmt/cbor"
+	atlas "github.com/polydawn/refmt/obj/atlas"
+	cbg "github.com/whyrusleeping/cbor-gen"
+)
+
+type IpldStore interface {
```
Comment (Stebalien) [inline store.go]: Docs? Lots of docs...

Comment (frrist) [reply store.go]: Added here: https://github.com/ipfs/go-ipld-cbor/pull/71

**Pull request merged**